### PR TITLE
JS: classify more HTML as generated

### PR DESCRIPTION
**Pull request by esben-semmle**
This change is motivated by the query `js/duplicate-html-id` being prone to flagging generated HTML.

The added heuristic for identifying generated HTML is that there are "many" HTML elements that start on the same line.

For now, I have selected `"many" > 20` as the limit, and I have not seen any FPs with that. See <https://lgtm.com/query/7721678073770499188/>, <https://lgtm.com/query/613928986248972020/> for exploratory use of this classifier.

Some examples of files that are classified by this change that are not classified otherwise:

-   <https://lgtm.com/projects/g/rapidpro/rapidpro/snapshot/417239731f0afc3f2017d6b8e8407cb427f91413/files/static/bower/jquery/speed/index.html?sort=name&dir=ASC&mode=list#L0>
-   <https://lgtm.com/projects/g/ckeditor/ckeditor-dev/snapshot/fb1ba41c8a4313ba2d9651b453b4c48b252fb502/files/plugins/magicline/dev/magicline.html?sort=name&dir=ASC&mode=list#L0>
-   <https://lgtm.com/projects/g/alibaba/ice/snapshot/aba04b6a5a5ee0c811f054294e46beef4e2e4e14/files/react-materials/scaffolds/ice-opensource-site/en-us/blog/blog1.html?sort=name&dir=ASC&mode=list#L0>
-   <https://lgtm.com/projects/g/Plateful/plateful-mobile/snapshot/5e819126d8882a5f5191e7b5af58452155b5a11e/files/coverage/Chrome%2036.0.1985%20>(Mac%20OS%20X%2010.9.4)/js/vendor.js.html?sort=name&dir=ASC&mode=list#L0
-   <https://lgtm.com/projects/g/capnmidnight/Primrose/snapshot/12920ddbe9f0c6a9f6076c63bc0bd9bb1840b4fb/files/demos/calendar.html?sort=name&dir=ASC&mode=list#L0>
-   <https://lgtm.com/projects/g/capnmidnight/Primrose/snapshot/12920ddbe9f0c6a9f6076c63bc0bd9bb1840b4fb/files/index.html?sort=name&dir=ASC&mode=list#L0>
-   <https://lgtm.com/projects/g/pomber/code-surfer/snapshot/d0cce0ec43f177a15c7d8d414ffac52bd5cef7fb/files/website/public/index.html?sort=name&dir=ASC&mode=list>
-   <https://lgtm.com/projects/g/Tencent/omi/snapshot/9ec05902485f1df96977236f3ee714ef508363e8/files/packages/omi-snap/build/index.html?sort=name&dir=ASC&mode=list#L0>
-   <https://lgtm.com/projects/g/ampproject/amphtml/snapshot/7bc3fed814940e6d9df6b4638f9a87a27c9a9cb5/files/spec/amp-cache-modifications.everything.cache.html?sort=name&dir=ASC&mode=list#L0>

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/GeneratedCode.qll [hunk 1]
```diff
@@ -121,6 +121,27 @@ private predicate isGeneratedHtml(File f) {
...
+}
+
+
+/**
+ * Gets the number of HTML elements that starts at line `l` in file `f`.
+ */
+private int countStartingHtmlElements(File f, int l) {
+  exists(HTML::Element candidate |
+    candidate.getFile() = f and
+    candidate.getLocation().getStartLine() = l and
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/GeneratedCode.qll]: Isn't that just `candidate = getAStartingElement(f, l)`?

File: javascript/ql/src/semmle/javascript/GeneratedCode.qll [hunk 2]
```diff
@@ -121,6 +121,27 @@ private predicate isGeneratedHtml(File f) {
...
+    result = count(getAStartingElement(f, l))
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/GeneratedCode.qll]: ...and isn't this whole predicate just `result = strictcount(getAStartingElement(f, l))`?

**Comment by esben-semmle** [comment]
right, I had forgotten about `strictcount`, that was silly.
Performance is unchanged for big-apps.slugs: https://git.semmle.com/esben/dist-compare-reports/tree/js/classify-mode-html_1553538000384

File: javascript/ql/src/semmle/javascript/GeneratedCode.qll [hunk 3]
```diff
@@ -121,6 +121,23 @@ private predicate isGeneratedHtml(File f) {
...
+/**
+ * Gets an element that starts at line `l` in file `f`.
+ */
+private HTML::Element getAStartingElement(File f, int l) {
+  result.getFile() = f and result.getLocation().getStartLine() = l
+}
+
+
+/**
+ * Gets the number of HTML elements that starts at line `l` in file `f`.
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/GeneratedCode.qll]:
```suggestion
 * Gets the number of HTML elements that start at line `l` in file `f`.
```

**Pull request merged**